### DISCUSSION TITLE
Capsule update with USBIO in FmpDxe.

### DISCUSSION THREAD

**Discussion by gordontcp** [tianocore/edk2#3143] [Q&A]
Hi all,
In `\FmpDevicePkg\Library\FmpDeviceLibNull\FmpDeviceLib.c`
There are comments for RegisterFmpInstaller:
```
/**
  Used to pass the FMP install function to this lib.  This allows the library to
  have control of the handle that the FMP instance is installed on.  This allows
  the library to use DriverBinding protocol model to locate its device(s) in the
  system.

  @param[in] Func  Function pointer to FMP install function.

  @retval EFI_SUCCESS       Library has saved function pointer and will call
                            function pointer on each DriverBinding Start.
  @retval EFI_UNSUPPORTED   Library doesn't use driver binding and only supports
                            a single instance.
  @retval other error       Error occurred.  Don't install FMP

**/
EFI_STATUS
EFIAPI
RegisterFmpInstaller (
  IN FMP_DEVICE_LIB_REGISTER_FMP_INSTALLER Func
  )
{
  //
  // This is a system firmware update that does not use Driver Binding Protocol
  //
  return EFI_UNSUPPORTED;
}
```
However, since the ‘RegisterFmpInstaller’ of the example directly returns `EFI_UNSUPPORTED`, it does not demonstrate how to handle the DriverBinding protocol.

In addition, I refer to
the example of `\MdeModulePkg\Bus\Usb\UsbMouseAbsolutePointerDxe` adding the driverbinding protocol to locate the USBMouse device to the 'RegisterFmpInstaller' function. And set the return value of RegisterFmpInstaller to `EFI_SUCCESS`.
where RegisterFmpInstaller is in the following file:
`edk2-platforms\Platform\Intel\Vlv2TbltDevicePkg\Feature\Capsule\Library\FmpDeviceLibSample`
Then tested on Minnowboard.
Test command: `CapsuleApp.efi Red.cap`
The test results are as follows:
1. During the Capsule update process, when `FmpDeviceSetImageWithStatus` is executed, `USBMouseAbsolutePointerDriverBindingSupported` has not been executed, that is, USB Device IO cannot be operated yet.
2. After `USBMouseAbsolutePointerDriverBindingSupporte`d is executed, USBIO can be operated, but at this time `FmpDeviceSetImageWithStatus` has been executed, that is, capsule update is too late to process.

My questions are as follows:
1. Is there an example of how RegisterFmpInstaller handles the driverbinding protocol?
2. When processing the Capsule update (FmpDeviceSetImageWithStatus), the USBMouseAbsolutePointerDriverBindingSupported has not been executed. What should I do?
3. The parameter in RegisterFmpInstaller is the function pointer '`FMP_DEVICE_LIB_REGISTER_FMP_INSTALLER Func`', and how should the parameter `Handle` in `Func` (ie `InstallFmpInstance`, see below) be passed from `FmpDxe` to `FmpDeviceLib`?
```
EFI_STATUS
EFIAPI
FmpDxeEntryPoint (
IN EFI_HANDLE ImageHandle,
IN EFI_SYSTEM_TABLE *SystemTable
)
{
….
Status = RegisterFmpInstaller(InstallFmpInstance);
…
}
EFI_STATUS
EFIAPI
InstallFmpInstance (
IN EFI_HANDLE Handle
)
{….
}
EFI_STATUS
EFIAPI
RegisterFmpInstaller(
IN FMP_DEVICE_LIB_REGISTER_FMP_INSTALLER Func
)
{
// Because this is a sample lib with very simple fake device we don't use
// the driverbinding protocol to locate our device.
//
return EFI_UNSUPPORTED;
}

```

**Comment by gordontcp** [comment]
Any suggestion is highly appreciated!
Thanks!

**Comment by gordontcp** [comment]
Question 3 is not a problem because the parameters are not passed in by FmpDxe

**Comment by spbrogan** [comment]
If i understand correctly you are suggesting that your FmpProtocol isn't yet installed by the time the BDS does the capsule update processing.  This is a choice and a design of the BDS.  In some BDS designs I have seen capsules processes twice.  Once in Dxe for known platform components and once again later for 3rd party or optionally connected devices.  

In cases where I have used UEFI drivers (drivers that use driver binding) I have added custom code to the BDS to connect that device prior to processing capsules.  

See https://github.com/tianocore/edk2/blob/0dc9b78a46813d61533b2bb0f7ef897a06a273be/MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleProcessLib.c#L676

Code calling this is very inconsistent, as is every BDS in tianocore (they are all different and don't represent what you usually will see in PC products).  See here for an example: https://cs.github.com/tianocore/edk2-platforms/blob/86759a1bf82904fec4ce5abe3bb7c04948822767/Platform/RaspberryPi/Library/PlatformBootManagerLib/PlatformBm.c
